metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -0,0 +1,285 @@
+        case Success(uri) => Right(uri)
+      }
+      _ <-
+        if (uri.getScheme != "metalsDecodeFile") Left(InvalidScheme(uri))
+        else Right(uri)
+      paramArr <- {
+        val paramArr = uri.getQuery.split("&").map(_.split("=", 2))
+        if (paramArr.exists(_.length != 2)) Left(InvalidQuery(uri))
+        else Right(paramArr)
+      }
+      params <- Try {
+        paramArr
+          .map(f => f.map(URLDecoder.decode(_, StandardCharsets.UTF_8)))
+          .map(f => f(0) -> f(1))
+          .toMap
+      } match {
+        case Failure(ex) => Left(InvalidQueryParams(paramArr, ex))
+        case Success(params) => Right(params)
+      }
+      _ <-
+        if (!params.contains("decoder")) Left(MissingDecoder(uri))
+        else Right(params)
+      _ <-
+        if (!params.contains("originalURI")) Left(MissingOriginalURI(uri))
+        else Right(params)
+      filePath <- {
+        val originalURIAsStr = params("originalURI")
+        Try {
+          originalURIAsStr.toAbsolutePath
+        } match {
+          case Failure(ex) => Left(InvalidURIPath(originalURIAsStr, ex))
+          case Success(filePath) => Right(filePath)
+        }
+      }
+      decoder <-
+        if (!filePath.exists) Left(FileMissing(filePath))
+        else getDecoder(params)
+      fileToDecode <- {
+        val fileToDecode = decoder.findFile(filePath)
+        if (fileToDecode.isEmpty) Left(FileToDecodeMissing(filePath))
+        else Right(fileToDecode.get)
+      }
+    } yield (decoder, fileToDecode)
+  }
+
+  private def getDecoder(
+      params: Map[String, String]
+  ): Either[Error, Decoder] = {
+    params("decoder") match {
+      case "javap" => Right(getJavapDecoder)
+      case "semanticdb" => Right(getSemanticdbDecoder)
+      case default => Left(InvalidDecoder(default))
+    }
+  }
+
+  private def getJavapDecoder: Decoder = {
+    Decoder(
+      path =>
+        if (path.isScalaOrJava)
+          findClassesDirFileFromSource(path, "class")
+        else Some(path),
+      decodeJavapFromClassFile
+    )
+  }
+
+  private def getSemanticdbDecoder: Decoder = {
+    Decoder(
+      path =>
+        if (path.isScalaOrJava)
+          findSemanticDBFileFromSource(path)
+        else Some(path),
+      decodeFromSemanticDBFile
+    )
+  }
+
+  private def findClassesDirFileFromSource(
+      sourceFile: AbsolutePath,
+      newExtension: String
+  ): Option[AbsolutePath] = {
+    for {
+      targetId <- buildTargets.sourceBuildTargets(sourceFile).headOption
+      target <- buildTargets.scalaTarget(targetId)
+      sourceRoot <- buildTargets.inverseSourceItem(sourceFile)
+      classDir = target.classDirectory.toAbsolutePath
+      oldExtension = sourceFile.extension
+      relativePath = sourceFile
+        .toRelative(sourceRoot)
+        .resolveSibling(_.stripSuffix(oldExtension) + newExtension)
+    } yield classDir.resolve(relativePath)
+  }
+  private def findSemanticDBFileFromSource(
```

Comment:
I'm wondering maybe we can re-use the `FileSystemSemanticdbs` to find the corresponding semanticdb file from the given file path. It looks like there's some duplication with the implementation in `FileSystemSemanticdbs`, and I think it would be great if we could use the same procedure.

https://github.com/scalameta/metals/blob/ef9cf6462867c1a56d06eb813f2889a864263e14/metals/src/main/scala/scala/meta/internal/metals/FileSystemSemanticdbs.scala#L57

But, this can be done in a separate issue/PR :)